### ISSUE TITLE
Do the extracted image features contain contextual information or only category information.

### ISSUE THREAD

**Issue by jkff00** [SuleBai/SC-CLIP#3]

**Comment by jkff00** [comment]
Is it ok if I want to get the caption for a specified area (like providing mask), but there are no preset categories.

**Comment by SuleBai** [comment]
Yes, you can refer to the `MaskPooling` code below. After obtaining the class-agnostic, pixel-aligned embeddings in the code [here](https://github.com/SuleBai/SC-CLIP/blob/74471da714dad9f565558a4501b0db7199fd0bf9/scclip_segmentor.py#L70), you can use the `MaskPooling` operation to extract the semantic embedding of the specified region using the provided mask.

```
class MaskPooling(nn.Module):
    def __init__(self,):
        super().__init__()

    def forward(self, x, mask):
        """
        Args:
            x: [B, C, H, W]
            mask: [B, Q, H, W]
        """
        if not x.shape[-2:] == mask.shape[-2:]:
            # reshape mask to x
            mask = F.interpolate(mask, size=x.shape[-2:], mode='bilinear', align_corners=False)
        with torch.no_grad():
            mask = mask.detach()
            mask = (mask > 0).to(mask.dtype)
            denorm = mask.sum(dim=(-1, -2), keepdim=True) + 1e-8

        mask_pooled_x = torch.einsum(
            "bchw,bqhw->bqc",
            x,
            mask / denorm,
        )
        return mask_pooled_x
```
